### ISSUE TITLE
Decide how to handle MDC / Logging in Akka Typed

### ISSUE THREAD

**Issue by ktoso** [akka/akka#23326]
Since people will be asking what is the Typed way to obtain a logger "inside an actor".

For the `Actor.immutable` style this may well mean - "you don't", just use whichever logging library you like?
Otherwise there may be a case to allow `Logging(typedsystem, source)` etc

Depends what we want to do with `Logging` in general, let's discuss in general and then decide what to document or implement.

**Comment by ktoso** [comment]
See a question on this on akka-user https://groups.google.com/forum/#!topic/akka-user/TmzMVgZjdnU

cc @rkuhn

**Comment by rkuhn** [comment]
The factory just picks up the ActorRef and some class information (if available) to be printed as source, and it is a type-class so it should be easily extensible, also for end-users. Hence I think this is mainly a question of adding a LogSource instance for typed ActorRef plus some docs.

**Comment by ktoso** [comment]
Right, the other question would be about a Behavior that perhaps decorates normal ones, and could be used to add / clear things from the MDC context. Should also be simple enough to implement actually.

**Comment by raboof** [comment]
This would be really nice - we'd need to give some guidance/recommendations how to make sure to include actor context information in the log and how to do `EventFilter`-style things in tests (even when running tests in parallel) - IME these are very useful features and not entirely trivial to achieve. An 'MDC decorator' might already go a long way indeed...

**Comment by patriknw** [comment]
Perhaps just add `log: LoggingAdapter` to `ActorContext`?

**Comment by raboof** [comment]
This means you cannot "just use whichever logging library you like", though, you'll have to go through `LoggingAdapter`.

I see 2 ways to allow using "whichever logging library/API you like" and still easily inject ActorRef information:
* `MDC`-style: we can make sure the MDC is filled with the appropriate metadata before calling the behavior, but that means logging from asynchronous sections would not work - seems dangerous
* Add a user-supplied `T` field to `ActorContext`. It looks like that would mean adding a new generic type parameter in a ton of places though.. not sure if it's worth it.

While "bring your own API/library" would be nice, I'm not sure we could do it in a way that would not lose useful control/information..

**Comment by dhpiggott** [comment]
FWIW just as an extra data point, I've been doing `val log = Logging(context.system.toUntyped, context.self.toUntyped)` in an `Actor.deferred` and then passing that to my actual behaviors. I went with this approach because I wanted to still be able to use `%X{akkaSource}` in my Logback encoder, and without reinventing the wheel.

Full example in case that's not clear: https://github.com/dhpcs/liquidity/blob/a9e588b84d0be0df359dd2ab9ef6039d58a9ae23/server/src/main/scala/com/dhpcs/liquidity/server/actor/ZoneMonitorActor.scala#L22.

**Comment by jrudolph** [comment]
> logging from asynchronous sections would not work

WDYM by that exactly and how would does it work in untyped?

I think there might be a third option, where you can configure an ActorSystem with code to execute on initialization and before and after actor execution with enough metadata to setup the context for logging.

**Comment by patriknw** [comment]
Note that this ticket is not only about MDC, it's also how to make it convenient to create a LoggingAdapter for typed actors. I'm tempted to simply add it to the `ActorContext` with lazy init. It has the small drawback that it takes up 4 bytes per actor even if logging is not used.
